internal/backend/sema/backend.go

```diff
@@ -0,0 +1,99 @@
+package sema
+
+import (
+	"context"
+	"io"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/restic/restic/internal/debug"
+	"github.com/restic/restic/internal/errors"
+	"github.com/restic/restic/internal/restic"
+)
+
+// make sure that SemaphoreBackend implements restic.Backend
+var _ restic.Backend = &SemaphoreBackend{}
+
+// SemaphoreBackend limits the number of concurrent operations.
+type SemaphoreBackend struct {
```

Comment:
Maybe LimitedBackend and rename the package semaphore?

Reply:
There is already a `rateLimitedBackend`, adding a LimitedBackend could be quite confusing in that regard.

Reply:
Fair enough.

Reply:
what about `connectionLimitedBackend` then? That would be somewhat more descriptive.